File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.java
import java.util.Scanner;

public class recurrence_analysis {
    private static void printIndent(int depth) {
        for (int i = 0; i < depth; i++) {
            System.out.print("  |  ");
        }
    }

    private static String getAddress(int depth) {
        long base = 0x7ffebfaf1000L;
        return "0x" + Long.toHexString(base - depth * 0x30L);
    }

    public static long calculateFactorial(int n, int depth) {
        printIndent(depth);
        System.out.println("Entering factorial(" + n + "). Address of n: " + getAddress(depth));
        if (n <= 1) {
            printIndent(depth);
            System.out.println("Base Case reached for factorial(" + n + "). Returning 1.");
            return 1;
        }
        long subResult = calculateFactorial(n - 1, depth + 1);
        long result = n * subResult;
        printIndent(depth);
        System.out.println("Exiting factorial(" + n + ") -> returning " + n + " * " + subResult + " = " + result);
        return result;
    }

    public static int calculateFibonacci(int n, int depth, String branch) {
        printIndent(depth);
        System.out.println("Entering fibonacci(" + n + ") [" + branch + " branch]. Address of n: " + getAddress(depth));
        if (n == 0) {
            printIndent(depth);
            System.out.println("Base Case: fibonacci(0) = 0. Returning.");
            return 0;
        }
        if (n == 1) {
            printIndent(depth);
            System.out.println("Base Case: fibonacci(1) = 1. Returning.");
            return 1;
        }
        int left = calculateFibonacci(n - 1, depth + 1, "Left");
        int right = calculateFibonacci(n - 2, depth + 1, "Right");
        int result = left + right;
        printIndent(depth);
        System.out.println("Exiting fibonacci(" + n + ") -> returning " + left + " + " + right + " = " + result);
        return result;
    }

    public static void main(String[] args) {
        int val = 5;
        System.out.println("=== Recursion & Stack Frame Visualizer ===\n");
        System.out.println("--- PART 1: Linear Recursion - Factorial(" + val + ") ---");
        long fact = calculateFactorial(val, 0);
        System.out.println("\nFactorial of " + val + " is: " + fact + "\n");
        System.out.println("--------------------------------------------------\n");
        System.out.println("--- PART 2: Tree Recursion - Fibonacci(4) ---");
        int fib = calculateFibonacci(4, 0, "Root");
        System.out.println("\nFibonacci of 4 is: " + fib + "\n");
        System.out.print("Press Enter to exit...");
        Scanner scanner = new Scanner(System.in);
        scanner.nextLine();
        scanner.close();
    }
}

File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.kt
import java.util.Scanner

fun printIndent(depth: Int) {
    for (i in 0 until depth) {
        print("  |  ")
    }
}

fun calculateFactorial(n: Int, depth: Int): Long {
    printIndent(depth)
    val address = "0x" + Integer.toHexString(System.identityHashCode(n))
    println("Entering factorial($n). Address of n: $address")
    if (n <= 1) {
        printIndent(depth)
        println("Base Case reached for factorial($n). Returning 1.")
        return 1
    }
    val subResult = calculateFactorial(n - 1, depth + 1)
    val result = n * subResult
    printIndent(depth)
    println("Exiting factorial($n) -> returning $n * $subResult = $result")
    return result
}

fun calculateFibonacci(n: Int, depth: Int, branch: String): Int {
    printIndent(depth)
    val address = "0x" + Integer.toHexString(System.identityHashCode(n))
    println("Entering fibonacci($n) [$branch branch]. Address of n: $address")
    if (n == 0) {
        printIndent(depth)
        println("Base Case: fibonacci(0) = 0. Returning.")
        return 0
    }
    if (n == 1) {
        printIndent(depth)
        println("Base Case: fibonacci(1) = 1. Returning.")
        return 1
    }
    val left = calculateFibonacci(n - 1, depth + 1, "Left")
    val right = calculateFibonacci(n - 2, depth + 1, "Right")
    val result = left + right
    printIndent(depth)
    println("Exiting fibonacci($n) -> returning $left + $right = $result")
    return result
}

fun main() {
    val valSize = 5
    println("=== Recursion & Stack Frame Visualizer ===\n")
    println("--- PART 1: Linear Recursion - Factorial($valSize) ---")
    val fact = calculateFactorial(valSize, 0)
    println("\nFactorial of $valSize is: $fact\n")
    println("--------------------------------------------------\n")
    println("--- PART 2: Tree Recursion - Fibonacci(4) ---")
    val fib = calculateFibonacci(4, 0, "Root")
    println("\nFibonacci of 4 is: $fib\n")
    print("Press Enter to exit...")
    Scanner(System.`in`).nextLine()
}
